src/mbr.c

```diff
@@ -0,0 +1,270 @@
+#include <errno.h>
+#include <fcntl.h>
+#include <glib/gstdio.h>
+#include <unistd.h>
+#include <sys/ioctl.h>
+#include <sys/mount.h>
+#include <linux/hdreg.h>
+
+#include "mbr.h"
+#include "update_handler.h"
+
+/* partition entry in MBR partition table, the system boots from */
+#define BOOT_PARTITION_ENTRY		0
+#define MBR_NUMBER_OF_PARTITIONS	4
+#define MBR_MAGIC_NUMBER_L		0x55
+#define MBR_MAGIC_NUMBER_H		0xAA
+
+#define DEFAULT_SECTOR_SIZE		512
+
+struct mbr {
+	guint8 bootstrap_code[440];
+	guint32 disk_signature;
+	guint16 unused;
+	struct partition_tbl_entry {
+		guint8 boot_indicator;
+		guint8 chs_begin[3];
+		guint8 type;
+		guint8 chs_end[3];
+		guint32 partition_start;
+		guint32 partition_size;
+	} partition_table[MBR_NUMBER_OF_PARTITIONS];
+	guint8 magic_number[2];
+} __attribute__((packed));
+
+static gboolean get_sectorsize(gint fd, guint *sector_size)
+{
+	if (ioctl(fd, BLKSSZGET, sector_size) != 0)
+		return FALSE;
+
+	return TRUE;
+}
+
+static gboolean get_hd_geometry(gint fd, struct hd_geometry *geometry,
+		GError **error)
+{
+	if (ioctl(fd, HDIO_GETGEO, geometry) != 0) {
```

Comment:
Do the usual ``g_return_val_if_fail(error == NULL || *error == NULL, FALSE);`` here and below as well. Also, pointers which are filled should be checked with g_assert_nonnull(pointer);